lib/SIL/Mangle.cpp

```diff
@@ -45,6 +45,12 @@ using namespace Mangle;
 //===----------------------------------------------------------------------===//
 
 static void mangleSubstitution(Mangler &M, Substitution Sub) {
+  if (Sub.getReplacement()->hasArchetype() ||
+      Sub.getReplacement()->hasTypeParameter()) {
```

Comment:
You should always have one or the other (archetype or type parameter) and never both. Pick one case, and add an assertion for the other.